src/client/websocket/handlers/TYPING_START.js

```diff
@@ -5,14 +5,63 @@ const { Events } = require('../../../util/Constants');
 module.exports = (client, { d: data }) => {
   const channel = client.channels.cache.get(data.channel_id);
   const user = client.users.cache.get(data.user_id);
+  const timestamp = new Date(data.timestamp * 1000);
 
   if (channel && user) {
-    /**
-     * Emitted whenever a user starts typing in a channel.
-     * @event Client#typingStart
-     * @param {Channel} channel The channel the user started typing in
-     * @param {User} user The user that started typing
-     */
-    client.emit(Events.TYPING_START, channel, user);
+    if (channel.type === 'voice') {
+      client.emit(Events.WARN, `Discord sent a typing packet to a voice channel ${channel.id}`);
+      return;
+    }
+
+    if (channel._typing.has(user.id)) {
+      const typing = channel._typing.get(user.id);
+
+      typing.lastTimestamp = timestamp;
+      typing.resetTimeout(tooLate(channel, user));
+    } else {
+      channel._typing.set(user.id, new TypingData(client, timestamp, timestamp, tooLate(channel, user)));
+
+      /**
+       * Emitted whenever a user starts typing in a channel.
+       * @event Client#typingStart
+       * @param {Channel} channel The channel the user started typing in
+       * @param {User} user The user that started typing
+       */
+      client.emit(Events.TYPING_START, channel, user);
+    }
   }
 };
+
+class TypingData {
+  constructor(client, since, lastTimestamp, _timeout) {
+    this.client = client;
+
+    this.since = since;
+
+    this.lastTimestamp = lastTimestamp;
+
+    this._timeout = _timeout;
+  }
+
+  resetTimeout(_timeout) {
+    this.client.clearTimeout(this._timeout);
+    this._timeout = _timeout;
+  }
+
+  get elapsedTime() {
+    return Date.now() - this.since - 3000;
+  }
+}
+
+function tooLate(channel, user) {
+  return channel.client.setTimeout(() => {
+    /**
+     * Emitted whenever a user stops typing in a channel.
+     * @event Client#typingStop
+     * @param {Channel} channel The channel the user was typing in.
+     * @param {User} user The user that was typing.
+     */
+    channel.client.emit(Events.TYPING_STOP, channel, user, channel._typing.get(user.id));
```

Comment:
the 3rd parameter here still needs adding to the jsdoc and typings

Reply:
Thank you for the continuous feedback, I really appreciate it. 👍 

I've update typings for the `Channel#_typing` property; here's a list of all the channel objects that have been updated:
- [x] `TextChannel`
- [x] `NewsChannel`
- [x] `DMChannel`

Please let me know if there's anything else that needs to be done.